src/node_crypto.cc

```diff
@@ -1140,6 +1181,46 @@ void SecureContext::LoadPKCS12(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+#ifndef OPENSSL_NO_ENGINE
+void SecureContext::SetClientCertEngine(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_EQ(args.Length(), 1);
+  CHECK(args[0]->IsString());
+
+  SecureContext* sc = Unwrap<SecureContext>(args.This());
+
+  MarkPopErrorOnReturn mark_pop_error_on_return;
+
+  // SSL_CTX_set_client_cert_engine does not itself 'support' multiple
+  // calls by cleaning up before overwriting the client_cert_engine
+  // internal context variable.
+  // Instead of trying to fix up this problem we in turn also do not
+  // support multiple calls to SetClientCertEngine. As this may surprise
+  // a developer we throw him an exception.
+  if (sc->ctx_->client_cert_engine != nullptr) {
+    return env->ThrowError(
+        "Multiple calls to SetClientCertEngine are not allowed");
+  }
+
+  const node::Utf8Value engine_id(env->isolate(), args[0]);
+  ENGINE* engine = LoadEngineById(env, *engine_id);
```

Comment:
Just a thought... It *could* be overkill, but wrapping this in a class that does the `ENGINE_free()` automatically when it falls out of scope may be cleaner... e.g. 

```C
Engine engine(env, *engine_id);
if (*engine == nullptr) { /* fail */ }
int r = SSL_CTX_set_client_cert_engine(sc->ctx, *engine); ...
```

Reply:
(basically exactly like `ClearErrorOnReturn` :-) ...)

Reply:
You could use a `std::unique_ptr` with a custom deleter for that: http://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr

OTOH, the init and fini are so close together here that it might indeed be overkill.